### DISCUSSION TITLE
create wrong number of messagebuffers(m_port_buffers) for switch

### DISCUSSION THREAD

**Discussion by gxflying** [gem5/gem5#583] [gem5-dev]
Hi sir, 
I am trying to figure out the connection of the ruby network by MessageBuffers recently
and when I dive into the code I found that maybe the code by now create wrong number of MessageBuffers for **Switch** in setup_buffers
https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/network/simple/SimpleNetwork.py#L104
https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/network/simple/SimpleNetwork.py#L152
```
        for link in network.int_links:
            if link.dst_node == self:
                for i in range(int(network.number_of_virtual_networks)):
                    router_buffers.append(
                        SwitchPortBuffer(buffer_size=vnet_buffer_size(i))
                    )
```
 because one mesage buffer  is created each time the switch **acts as a destionation node,** and the messageBuffers are consumed in Switch::addOutPort,  **BUT** the Switch::addOutPort is called only when the switch **acts as a source node**
this is not a problem when the topology of the network is a full connected pt2pt network
but when the connection of the nodes in the network is NOT full connected , it will be a problem

as a example with 3 nodes,  the Node0 acts src-mode two times but acts as dest-node one time
when run with the current gem5 code , there will be a **accessing out-of-bounds** of the m_port_buffer array

> [0, 1]  // meaning for uni-direction from src :Node0 to dest Node1
> [0, 2]
> [1,2]
> [2,0]
> [2,1]

https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/network/simple/Switch.cc#L88
```
    for (int i = 0; i < out.size(); ++i) {
        assert(m_num_connected_buffers < m_port_buffers.size());
        MessageBuffer* buffer_ptr =
            m_port_buffers[m_num_connected_buffers];
        m_num_connected_buffers++;
        intermediateBuffers.push_back(buffer_ptr);
    }
```
https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/network/simple/SimpleNetwork.cc#L122
https://github.com/gem5/gem5/blob/d0d3c74ce00f6d852a39060f3b7ce083887b1cfd/src/mem/ruby/network/simple/SimpleNetwork.cc#L149

**Comment by MeatBoy106** [comment]
Your observation seems correct. I think this can be reported as a bug.
